### PR TITLE
add go module support to copilot

### PR DESCRIPTION
**Pull request by zachgersh**
@cloudfoundry/cf-routing !

Go module support has now made its way to copilot and we've learned quite a bit PR'ing this to other CF repos 💃 

## Some helpful bits:

1) When pulling this in your scripts that test this in CI will now need to run `go mod download` before starting to run tests or build. If you like to discuss why please ping me!
2) You can now build copilot OUTSIDE of the `GOPATH` - feel free to clone this project wherever you want!
3) If building inside the `GOPATH` you need to turn module support on (not recommended).
4) You can totally build a `go mod` project inside of your `GOPATH` even though it is not recommend (see discussion below) by setting this: `GO111MODULE=on`

When running ginkgo tests - if you do not pre-download your assets with the command mentioned in point 1 above, you may run into strange errors. This is due to the fact that ginkgo compiles all tests in parallel AND there is currently a bug (solved in 1.12) related to download to the go mod cache in parallel.

## In summary

Believe it is better to integrate early and often. Golang 1.12 will have more module support (shipping shortly) and 1.13 is when it becomes the default. Hoping to get every project ahead of the curve here.

## Additional Help

P.S. bonus tip - if you need to clear your mod cache so you can re-download assets use this helpful command `go clean -modcache` do not just delete the mod cache directory in `$GOPATH/pkg/mod`

P.P.S - if you'd like to dig into modules more, here are some resources

[the wiki](https://github.com/golang/go/wiki/Modules)
[go modules by example](https://github.com/go-modules-by-example/index)

### CODE REVIEW

**Comment by utako** [comment]
Do you have any concerns about using this before it'd been adopted for a little while, letting other people discover bugs and issues before pulling this in and potentially disrupting work in flight? I worry that without some caution, we'll all be running into the same issues, leading to some duplicated work to solve them.

The Routing and Container Networking teams would both likely want to learn about the hiccups you've experienced in this PR process as well, as those hiccups are likely things we'd experience if we were doing the change ourselves. Any additional context would be welcome.

**Comment by rosenhouse** [comment]
> If building inside the GOPATH you need to turn module support on (not recommended).

I'm concerned that the "not recommended" approach is what we'd be doing on a daily basis, given that Copilot is built and used exclusively inside of istio-release, which is a GOPATH.

**Comment by zachgersh** [comment]
@utako - I don't foresee too many more problems other than the ones I called out above. I could have also written a section about how do I add / remove / upgrade dependencies but I think we've taken care of the major stumbling blocks or issues around what it looks like to work with go modules. I'd be glad to discuss it more and hopefully I could address any other concerns around the process? I am also fully aware that there are going to be unknown unkowns so your point is well taken.

@rosenhouse - I say "not recommend" only because this is where the go community / maintainers are going. We are rapidly going to move toward a world where the `GOPATH` is required for building projects (this is great for everyone). 

You can totally build and work with a `go mod` based project in your `GOPATH` just be setting this explicitly: `GO111MODULE=on`

I know it takes awhile to change hearts and minds so getting used to a newer workflow is something we should start on somewhat soonish as it will take awhile to break old habits.

**Comment by utako** [comment]
As an aside: after a brief conversation at standup, we've decided to talk about this more in-depth as a team. I've added it as a topic during tech forum.

**Comment by zachgersh** [comment]
should have cc'ed @cloudfoundry/cf-container-networking as well (which @utako reminded me of above).

**Comment by zachgersh** [comment]
side note - I am going to fix this PR to solve the above conflict. I'd love if we could decide whether to pursue this or not in the next couple of weeks (just so I don't have to continue to fix the conflicts that come with having this open).

I need to come back through and revise the description of this as well.

**Comment by christianang** [comment]
We had a discussion about this and the consensus was that we would revisit switching to go modules once it is closer to being default in Go, maybe a few months from now. That being said, I wouldn't worry about any conflicts at the moment.
